### PR TITLE
Add highlighter

### PR DESCRIPTION
**Pull request by yasabere**
[card](https://rentpath.leankit.com/card/587790554)

component which adds class to text based on pattern prop

### CODE REVIEW

File: packages/react-ui-core/src/Highlighter/Highlighter.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      curr])
+  }
+
+  renderHighlightedSegment(index) {
+    const { pattern, theme, indexHighlighted } = this.props
+
+    if (indexHighlighted === undefined || index === indexHighlighted) {
+      return (
+        <span
+          data-tid="Highlighter-match"
```
Comment (tadjohnston) [inline packages/react-ui-core/src/Highlighter/Highlighter.js]: Always has to be lower cased 😃

File: packages/react-ui-core/src/Highlighter/Highlighter.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    return segments.reduce((prev, curr, index) => [prev,
+      this.renderHighlightedSegment(index),
+      curr])
+  }
+
+  renderHighlightedSegment(index) {
+    const { pattern, theme, indexHighlighted } = this.props
+
+    if (indexHighlighted === undefined || index === indexHighlighted) {
```
Comment (tadjohnston) [inline packages/react-ui-core/src/Highlighter/Highlighter.js]: What happens if `indexHighlighted` is null?

File: packages/react-ui-core/src/Highlighter/Highlighter.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    }
+    return pattern
+  }
+
+  render() {
+    const { theme } = this.props
+
+    return (
+      <div
+        className={classnames(theme.Highlighter)}
```
Comment (tadjohnston) [inline packages/react-ui-core/src/Highlighter/Highlighter.js]: No need for `classnames` here as you only have one class. I'm guessing you're missing `className` prop here which means you'd need `classnames` after all 😄

**Comment by tadjohnston** [comment]
Let's make sure we get some description in the card so we know what this card is really supposed to be doing.

File: packages/react-ui-core/src/Highlighter/Highlighter.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  get indexHighlighted() {
+    return this.props.indexHighlighted + 1
+  }
+
+  renderHighlightedText() {
+    const { pattern, children } = this.props
+
+    const segments = children.split(pattern)
```
Comment (tadjohnston) [inline packages/react-ui-core/src/Highlighter/Highlighter.js]: Should this return if no pattern exists? If we're forcing this to have a pattern, then it should be required in props.

File: packages/react-ui-core/src/Highlighter/Highlighter.js (hunk 2 above)
Comment (nray621) [reply packages/react-ui-core/src/Highlighter/Highlighter.js]: The propTypes enforce it as a number, so as far as I understand, it could only be `undefined` or a number at this point.

File: packages/react-ui-core/src/Highlighter/Highlighter.js (hunk 4 above)
Comment (nray621) [reply packages/react-ui-core/src/Highlighter/Highlighter.js]: 👍

File: packages/react-ui-core/src/Highlighter/__tests__/Highlighter-test.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+describe('Highlighter', () => {
+  it('surounds pattern text with span and classname', () => {
+    const wrapper = mount(
+      <Highlighter pattern="test">
+        This is a test sentence for a test
+      </Highlighter>
+    )
+
+    wrapper.setProps({ pattern: 'test' })
+    expect(wrapper.find('[data-tid="highlighter-match"]')).toHaveLength(2)
```
Comment (tadjohnston) [inline packages/react-ui-core/src/Highlighter/__tests__/Highlighter-test.js]: Let's set a const to this value `wrapper.find('[data-tid="highlighter-match"]')` instead of looking it up multiple times.  It also makes it easier to change if for some reason the data-tid changes.

File: stories/core/Highlighter.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    <Highlighter
+      className={StoryBookTheme['Story-padding']}
+      pattern="test"
+    >
+      this is a test to see if this works, test seems to have passed the test
+    </Highlighter>
+  </div>
+)
+
+export const HighlighterWithIndex = () => {
```
Comment (tadjohnston) [inline stories/core/Highlighter.js]: This story is broken.

File: packages/react-ui-core/src/Highlighter/__tests__/Highlighter-test.js (hunk 5 above)
Comment (nray621) [reply packages/react-ui-core/src/Highlighter/__tests__/Highlighter-test.js]: Sounds good 👍

File: packages/react-ui-core/src/Highlighter/__tests__/Highlighter-test.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    const wrapper = mount(
+      <Highlighter pattern="test" indexHighlighted={2}>
+        This is a test sentence for a test
+      </Highlighter>
+    )
+
+    wrapper.setProps({ pattern: 'test' })
+    expect(wrapper.find('[data-tid="highlighter-match"]')).toHaveLength(1)
+    wrapper.setProps({ pattern: 'test2' })
+    expect(wrapper.find('[data-tid="highlighter-match"]')).toHaveLength(0)
```
Comment (nray621) [inline packages/react-ui-core/src/Highlighter/__tests__/Highlighter-test.js]: I set a const for  `wrapper.find('[data-tid="highlighter-match"]')`, but had to keep line 19 as is. I think it needs to run `find` again after the prop change on line 18

File: packages/react-ui-core/src/Highlighter/Highlighter.js (hunk 2 above)
Comment (tadjohnston) [reply packages/react-ui-core/src/Highlighter/Highlighter.js]: If that's the case, can you add a test for `null`?

File: packages/react-ui-core/src/Highlighter/__tests__/Highlighter-test.js (hunk 7 above)
Comment (tadjohnston) [reply packages/react-ui-core/src/Highlighter/__tests__/Highlighter-test.js]: Most likely, yes.

File: stories/core/Highlighter.js (hunk 6 above)
Comment (nray621) [reply stories/core/Highlighter.js]: Let me take a look at why the knobs aren't working

Comment (nray621) [reply stories/core/Highlighter.js]: Fixed now

File: packages/react-ui-core/src/Highlighter/Highlighter.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  get indexHighlighted() {
+    return this.props.indexHighlighted + 1
+  }
+
+  renderHighlightedText() {
+    const { pattern, children } = this.props
+
+    if (pattern === '') return children
```
Comment (tadjohnston) [inline packages/react-ui-core/src/Highlighter/Highlighter.js]: I think you just want `if (!pattern) return children` yes?

Comment (nray621) [reply packages/react-ui-core/src/Highlighter/Highlighter.js]: duh. whoops

File: packages/react-ui-core/src/Highlighter/Highlighter.js [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    }
+    return pattern
+  }
+
+  render() {
+    const {
+      theme,
+      className,
+      indexHighlighted,
+      ...props
```
Comment (tadjohnston) [inline packages/react-ui-core/src/Highlighter/Highlighter.js]: `pattern` needs to be destuctured here as well I believe.

File: packages/react-ui-core/src/Highlighter/__tests__/Highlighter-test.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
+import React from 'react'
+import { mount } from 'enzyme'
+import Highlighter from '../Highlighter'
+
+describe('Highlighter', () => {
+  it('surounds pattern text with span and classname', () => {
+    const wrapper = mount(
```
Comment (tadjohnston) [inline packages/react-ui-core/src/Highlighter/__tests__/Highlighter-test.js]: I don't think we need to `mount` do we?

File: packages/react-ui-core/src/Highlighter/Highlighter.js (hunk 9 above)
Comment (nray621) [reply packages/react-ui-core/src/Highlighter/Highlighter.js]: 👍

File: packages/react-ui-core/src/Highlighter/__tests__/Highlighter-test.js (hunk 10 above)
Comment (nray621) [reply packages/react-ui-core/src/Highlighter/__tests__/Highlighter-test.js]: We'll check

File: packages/react-ui-core/src/AutoSuggestField/AutoSuggestField.js [hunk 11]
```diff
@@ -99,12 +102,17 @@ export default class AutoSuggestField extends Component {
   }
 
   get highlightedListItem() {
-    if (!this.props.highlight) return { nodeType: 'div' }
+    const { highlight } = this.props
+
+    if (!highlight) return { nodeType: 'div' }
+    let indexHighlighted = highlight.indexHighlighted || 1
+
+    if (highlight.indexHighlighted === 0) indexHighlighted = 0
```
Comment (tadjohnston) [inline packages/react-ui-core/src/AutoSuggestField/AutoSuggestField.js]: Why not just:
```
get highlightedListItem() {
  const { highlight } = this.props 

  if (!highlight) return { nodeType: 'div' }

  const { indexHighlighted: index, ...highlightProps } = highlight
  const indexHighlighted = index >= 0 ? index : 1

  return { 
    nodeType: props => (
      <Highlighter
        indexHighlighted={indexHighlighted}
        {...props} // not sure what this is btw
        {...highlightProps}
        pattern={this.state.value}
      />
    ),
  }
}

File: packages/react-ui-core/src/AutoSuggestField/AutoSuggestField.js [hunk 12]
```diff
@@ -95,6 +101,22 @@ export default class AutoSuggestField extends Component {
...
+    return { nodeType: props =>
+      (<Highlighter
+        indexHighlighted={indexHighlighted}
+        {...props}
```
[the remark this answers is not in the document]
Comment (nray621) [reply packages/react-ui-core/src/AutoSuggestField/AutoSuggestField.js]: On line 111?

File: packages/react-ui-core/src/Highlighter/__tests__/Highlighter-test.js (hunk 10 above)
Comment (yasabere) [reply packages/react-ui-core/src/Highlighter/__tests__/Highlighter-test.js]: These tests don't work with shallow

File: stories/core/List.js [hunk 13]
```diff
@@ -34,16 +34,15 @@ export const ListWithPassedNodeTypes = (
...
     }}
   />
-
 )
 
 export const ListWithOwnItemComponent = (
   <List
     items={sampleListItems}
-    ListItem={props => (<SampleItemComponent {...props} />)}
+    listItem={props => (<SampleItemComponent {...props} />)}
```
Comment (tadjohnston) [inline stories/core/List.js]: 👍

File: stories/theme/core/Highlighter.css [hunk 14]
```diff
@@ -0,0 +1,8 @@
+.Highlighter {
+
+}
+
+.Highlighter_Match {
```
Comment (tadjohnston) [inline stories/theme/core/Highlighter.css]: `Match` is a modifier so it should be `.Highlighter-match`

File: stories/theme/core/AutoSuggestField.css [hunk 15]
```diff
@@ -27,7 +27,15 @@
 }
 
 .AutoSuggestField .ListItem {
-  padding: 5px;
+  padding: 5px 0;
+}
+
+.AutoSuggestField .Highlighter_Match {
```
Comment (tadjohnston) [inline stories/theme/core/AutoSuggestField.css]: `Match` is a modifier so it should be `.Highlighter-match`

File: packages/react-ui-core/src/Highlighter/Highlighter.js [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+      indexHighlighted,
+      className,
+    } = this.props
+
+    if (!indexHighlighted || index === indexHighlighted) {
+      return (
+        <span
+          data-tid="highlighter-match"
+          className={classnames(
+            theme.Highlighter_Match,
```
Comment (tadjohnston) [inline packages/react-ui-core/src/Highlighter/Highlighter.js]: `Match` is a modifier so it should be `.Highlighter-match`

File: stories/theme/core/Highlighter.css (hunk 14 above)
Comment (nray621) [reply stories/theme/core/Highlighter.css]: Gotcha. Changing now

File: packages/react-ui-core/src/Highlighter/Highlighter.js [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+      pattern,
+      theme,
+      indexHighlighted,
+      className,
+    } = this.props
+
+    if (!indexHighlighted || index === indexHighlighted) {
+      return (
+        <span
+          data-tid="highlighter-match"
```
Comment (tadjohnston) [inline packages/react-ui-core/src/Highlighter/Highlighter.js]: Since this will be repeated, we may want to make the data-tid specific, but I'm not sure. Maybe:
```
data-tid=`highlighter-match-${index}`
```
I'm not sure if we _really_ need this, so please give thoughts on this.

Comment (nray621) [reply packages/react-ui-core/src/Highlighter/Highlighter.js]: @yasabere You a little more familiar with this code. What do you think?

File: packages/react-ui-core/src/Highlighter/Highlighter.js (hunk 16 above)
Comment (yasabere) [reply packages/react-ui-core/src/Highlighter/Highlighter.js]: Done

File: packages/react-ui-core/src/index.js [hunk 18]
```diff
@@ -48,6 +48,7 @@ export {
 
 export {
   List,
+  HighlightedList,
```
Comment (tadjohnston) [inline packages/react-ui-core/src/index.js]: This is not a component anymore is it?

Comment (nray621) [reply packages/react-ui-core/src/index.js]: Whoops. Sorry. I'll remove that

Comment (tadjohnston) [reply packages/react-ui-core/src/index.js]: Yeah, we need to be careful here. I think this would actually break things outside of this repo.

Comment (nray621) [reply packages/react-ui-core/src/index.js]: Gotcha. yeah, I'll be more careful with that going forward. That was my bad
